File: packages/typescript/src/context-summarizer.ts
/**
 * Context Summarizer — compresses large context histories to prevent token overflow.
 */

import type { ContextEntry, SharedContext } from "./types.js";

export interface SummarizerConfig {
  /** Maximum number of full entries to keep (most recent). Default: 3 */
  maxFullEntries: number;
  /** Maximum summary length for compressed entries. Default: 200 */
  maxSummaryLength: number;
  /** Maximum total context entries before summarization triggers. Default: 5 */
  triggerThreshold: number;
}

const DEFAULT_CONFIG: SummarizerConfig = {
  maxFullEntries: 3,
  maxSummaryLength: 200,
  triggerThreshold: 5,
};

/**
 * Summarize a context entry by truncating its summary.
 */
function compressEntry(entry: ContextEntry, maxLength: number): ContextEntry {
  return {
    ...entry,
    summary:
      entry.summary.length > maxLength
        ? entry.summary.slice(0, maxLength) + "..."
        : entry.summary,
    action: `[Summarized] ${entry.action}`,
  };
}

/**
 * Summarize the shared context history to reduce token usage.
 *
 * Keeps the most recent entries in full and compresses older ones.
 */
export function summarizeContext(
  context: SharedContext,
  config: Partial<SummarizerConfig> = {}
): SharedContext {
  const cfg = { ...DEFAULT_CONFIG, ...config };

  if (context.history.length <= cfg.triggerThreshold) {
    return context;
  }

  const fullEntries = context.history.slice(-cfg.maxFullEntries);
  const olderEntries = context.history.slice(
    0,
    context.history.length - cfg.maxFullEntries
  );

  const compressed = olderEntries.map((entry) =>
    compressEntry(entry, cfg.maxSummaryLength)
  );

  return {
    ...context,
    history: [...compressed, ...fullEntries],
  };
}

/**
 * Get an estimated token count for the context.
 * Uses a rough heuristic of ~4 characters per token.
 */
export function estimateContextTokens(context: SharedContext): number {
  let chars = 0;
  for (const entry of context.history) {
    chars += entry.role.length;
    chars += entry.action.length;
    chars += entry.summary.length;
    chars += entry.timestamp.length;
    if (entry.filesChanged) {
      chars += entry.filesChanged.join(",").length;
    }
  }
  for (const [path, content] of Object.entries(context.files)) {
    chars += path.length + content.length;
  }
  chars += JSON.stringify(context.metadata).length;
  return Math.ceil(chars / 4);
}

File: packages/python/src/tcs_bpi_ai_team/context_summarizer.py
"""Context Summarizer — compresses large context histories to prevent token overflow."""

from __future__ import annotations

import json
from dataclasses import dataclass

from tcs_bpi_ai_team.types import ContextEntry, SharedContext


@dataclass
class SummarizerConfig:
    """Configuration for context summarization."""

    max_full_entries: int = 3
    """Maximum number of full entries to keep (most recent)."""

    max_summary_length: int = 200
    """Maximum summary length for compressed entries."""

    trigger_threshold: int = 5
    """Maximum total context entries before summarization triggers."""


def _compress_entry(entry: ContextEntry, max_length: int) -> ContextEntry:
    """Summarize a context entry by truncating its summary."""
    summary = entry.summary
    if len(summary) > max_length:
        summary = summary[:max_length] + "..."

    return ContextEntry(
        role=entry.role,
        action=f"[Summarized] {entry.action}",
        summary=summary,
        timestamp=entry.timestamp,
        files_changed=entry.files_changed,
    )


def summarize_context(
    context: SharedContext,
    config: SummarizerConfig | None = None,
) -> SharedContext:
    """Summarize the shared context history to reduce token usage.

    Keeps the most recent entries in full and compresses older ones.
    """
    cfg = config or SummarizerConfig()

    if len(context.history) <= cfg.trigger_threshold:
        return context

    full_entries = context.history[-cfg.max_full_entries:]
    older_entries = context.history[: len(context.history) - cfg.max_full_entries]

    compressed = [_compress_entry(e, cfg.max_summary_length) for e in older_entries]

    return SharedContext(
        task_id=context.task_id,
        history=[*compressed, *full_entries],
        files=dict(context.files),
        metadata=dict(context.metadata),
    )


def estimate_context_tokens(context: SharedContext) -> int:
    """Estimate the token count for a context.

    Uses a rough heuristic of ~4 characters per token.
    """
    chars = 0
    for entry in context.history:
        chars += len(entry.role)
        chars += len(entry.action)
        chars += len(entry.summary)
        chars += len(entry.timestamp)
        if entry.files_changed:
            chars += len(",".join(entry.files_changed))

    for path, content in context.files.items():
        chars += len(path) + len(content)

    chars += len(json.dumps(context.metadata))
    return (chars + 3) // 4  # ceil division
